File: packages/js/src/libsql/crud.ts
import { CrudianError } from "../index.js"
import {
  createAsyncSqliteCrud,
  type AsyncSqliteCrud,
} from "../sqlite/async-crud.js"
import type { Row } from "../types.js"

/** Result shape used by the libSQL executor bridge. */
export type LibsqlResultSet = {
  rows: Array<Row | unknown>
  rowsAffected: number
}

/** Minimal execute surface shared by Client and Transaction. */
export type LibsqlExecutor = {
  execute(
    stmt: { sql: string; args?: unknown[] } | string,
  ): Promise<LibsqlResultSet>
}

/** Minimal surface of @libsql/client Client used by the adapter. */
export type LibsqlLikeClient = LibsqlExecutor & {
  transaction(
    mode?: "write" | "read" | "deferred",
  ): Promise<LibsqlLikeTransaction>
}

export type LibsqlLikeTransaction = LibsqlExecutor & {
  commit(): Promise<void>
  rollback(): Promise<void>
  close(): void
}

export type LibsqlCrud = AsyncSqliteCrud<LibsqlLikeClient>

function normalizeRow(row: unknown): Row {
  if (row == null || typeof row !== "object") {
    throw new CrudianError("expected row object")
  }
  const out: Row = { ...(row as Row) }
  for (const [k, v] of Object.entries(out)) {
    if (typeof v === "bigint") out[k] = Number(v)
  }
  return out
}

/**
 * Create a Crud bound to a libSQL client (`@libsql/client`).
 * Methods are async. The injected client is exposed as `crud.db`.
 * Auth / URL are configured on the caller-created client (e.g. from env).
 */
export function createCrud(client: LibsqlLikeClient): LibsqlCrud {
  if (client == null) {
    throw new CrudianError("db is required")
  }
  if (
    typeof client !== "object" ||
    typeof client.execute !== "function" ||
    typeof client.transaction !== "function"
  ) {
    throw new CrudianError("db must be a libSQL client")
  }

  let active: LibsqlExecutor = client

  return createAsyncSqliteCrud(client, {
    async run(sql, args = []) {
      const result = await active.execute({ sql, args })
      return { changes: Number(result.rowsAffected ?? 0) }
    },
    async get(sql, args = []) {
      const result = await active.execute({ sql, args })
      const row = result.rows[0]
      return row == null ? undefined : normalizeRow(row)
    },
    async all(sql, args = []) {
      const result = await active.execute({ sql, args })
      return result.rows.map((row) => normalizeRow(row))
    },
    async transaction<T>(fn: () => Promise<T>): Promise<T> {
      const tx = await client.transaction("write")
      const prev = active
      active = tx
      try {
        const value = await fn()
        await tx.commit()
        return value
      } catch (err) {
        await tx.rollback()
        throw err
      } finally {
        active = prev
        tx.close()
      }
    },
  })
}

File: packages/go/libsql/crud.go
package libsql

import (
	"context"
	"database/sql"

	"github.com/b4moss/crudian/go/crudian"
)

// Crud wraps *sql.DB (libSQL driver) with the shared CRUD facade.
type Crud struct {
	*crudian.Crud
	DB *sql.DB
}

// CreateCrud injects an existing *sql.DB. Does not open connections.
func CreateCrud(db *sql.DB) (*Crud, error) {
	if db == nil {
		return nil, crudian.NewError("db is required")
	}
	ex := &sqlExecutor{db: db}
	return &Crud{Crud: crudian.NewCrud(ex, crudian.SqliteDialect{}), DB: db}, nil
}

type sqlExecutor struct {
	db *sql.DB
	tx *sql.Tx
}

func (e *sqlExecutor) conn(ctx context.Context) querier {
	if e.tx != nil {
		return e.tx
	}
	return contextDB{db: e.db, ctx: ctx}
}

type querier interface {
	ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error)
	QueryContext(ctx context.Context, query string, args ...any) (*sql.Rows, error)
}

type contextDB struct {
	db  *sql.DB
	ctx context.Context
}

func (c contextDB) ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error) {
	if ctx == nil {
		ctx = c.ctx
	}
	return c.db.ExecContext(ctx, query, args...)
}

func (c contextDB) QueryContext(ctx context.Context, query string, args ...any) (*sql.Rows, error) {
	if ctx == nil {
		ctx = c.ctx
	}
	return c.db.QueryContext(ctx, query, args...)
}

func (e *sqlExecutor) Run(ctx context.Context, query string, args ...any) (int64, error) {
	res, err := e.conn(ctx).ExecContext(ctx, query, args...)
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}

func (e *sqlExecutor) Get(ctx context.Context, query string, args ...any) (crudian.Row, error) {
	rows, err := e.conn(ctx).QueryContext(ctx, query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	if !rows.Next() {
		return nil, nil
	}
	return scanRow(rows)
}

func (e *sqlExecutor) All(ctx context.Context, query string, args ...any) ([]crudian.Row, error) {
	rows, err := e.conn(ctx).QueryContext(ctx, query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]crudian.Row, 0)
	for rows.Next() {
		row, err := scanRow(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, row)
	}
	return out, rows.Err()
}

func (e *sqlExecutor) Transaction(ctx context.Context, fn func(tx crudian.Executor) error) error {
	tx, err := e.db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	inner := &sqlExecutor{db: e.db, tx: tx}
	if err := fn(inner); err != nil {
		_ = tx.Rollback()
		return err
	}
	return tx.Commit()
}

func scanRow(rows *sql.Rows) (crudian.Row, error) {
	cols, err := rows.Columns()
	if err != nil {
		return nil, err
	}
	raw := make([]any, len(cols))
	ptrs := make([]any, len(cols))
	for i := range raw {
		ptrs[i] = &raw[i]
	}
	if err := rows.Scan(ptrs...); err != nil {
		return nil, err
	}
	out := crudian.Row{}
	for i, c := range cols {
		v := raw[i]
		if b, ok := v.([]byte); ok {
			out[c] = string(b)
		} else {
			out[c] = v
		}
	}
	return out, nil
}
